open_spiel/games/solitaire.h

```diff
@@ -0,0 +1,409 @@
+
+// kMove actions are ones that are taken at decision nodes; they involve moving
+// a card to another cards location. It starts at 53 because there are 52 reveal
+// actions before it. See `NumDistinctActions()` in solitaire.cc.
+inline constexpr int kMoveStart = 53;
+inline constexpr int kMoveEnd = 204;
+
+// Indices for special cards
+inline constexpr int kHiddenCard = 99;
+inline constexpr int kEmptySpadeCard = -5;
+inline constexpr int kEmptyHeartCard = -4;
+inline constexpr int kEmptyClubCard = -3;
+inline constexpr int kEmptyDiamondCard = -2;
+inline constexpr int kEmptyTableauCard = -1;
+
+// 1 empty + 13 ranks
+inline constexpr int kFoundationTensorLength = 14;
+
+// 6 hidden cards + 1 empty tableau + 52 ordinary cards
+inline constexpr int kTableauTensorLength = 59;
+
+// 1 hidden card + 52 ordinary cards
+inline constexpr int kWasteTensorLength = 53;
+
+// Constant for how many hidden cards can show up in a tableau. As hidden cards
+// can't be added, the max is the highest number in a tableau at the start of
+// the game: 6
+inline constexpr int kMaxHiddenCard = 6;
+
+// Only used in one place and just for consistency (to match kChancePlayerId &
+// kTerminalPlayerId)
+inline constexpr int kPlayerId = 0;
+
+// Indicates the last index before the first player action (the last kReveal
+// action has an ID of 52)
+inline constexpr int kActionOffset = 52;
+
+// Order of suits
+const std::vector<SuitType> kSuits = {kSuitSpades, kSuitHearts, kSuitClubs,
+                                      kSuitDiamonds};
+
+// These correspond with their enums, not with the two vectors directly above
+const std::vector<std::string> kSuitStrs = {
+    "", kGlyphSpades, kGlyphHearts, kGlyphClubs, kGlyphDiamonds, ""};
+const std::vector<std::string> kRankStrs = {
+    "", "A", "2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", ""};
+
+const std::map<RankType, double> kFoundationPoints = {
+    // region Maps a RankType to the reward for moving a card of that rank to
+    // the foundation
+    {kRankA, 100.0}, {kRank2, 90.0}, {kRank3, 80.0}, {kRank4, 70.0},
+    {kRank5, 60.0},  {kRank6, 50.0}, {kRank7, 40.0}, {kRank8, 30.0},
+    {kRank9, 20.0},  {kRankT, 10.0}, {kRankJ, 10.0}, {kRankQ, 10.0},
+    {kRankK, 10.0}
+    // endregion
+};
+
+const std::map<SuitType, PileID> kSuitToPile = {
+    // region Maps a foundation suit to the ID of the foundation
+    {kSuitSpades, kPileSpades},
+    {kSuitHearts, kPileHearts},
+    {kSuitClubs, kPileClubs},
+    {kSuitDiamonds, kPileDiamonds}
+    // endregion
+};
+
+const std::map<int, PileID> kIntToPile = {
+    // region Maps an integer to a tableau pile ID (used when initializing
+    // SolitaireState)
+    {1, kPile1stTableau}, {2, kPile2ndTableau}, {3, kPile3rdTableau},
+    {4, kPile4thTableau}, {5, kPile5thTableau}, {6, kPile6thTableau},
+    {7, kPile7thTableau}
+    // endregion
+};
+
+// Support Classes =============================================================
```

Comment:
Do these need to be in the header file or can we put them in the .cc in an anonymous namespace?

Reply:
The declaration of SolitaireState depends on all of these classes, so unless I'm missing something, they can't be put in solitaire.cc. I can put them in a separate file and include them like chess/go/bridge if you want.